File: python/105.py
"""Project Euler Problem 105: Special Sum Sets.

Let S(A) represent the sum of elements in set A of size n. We shall call it
a special sum set if for any two non-empty disjoint subsets, B and C, the
following properties are true:
1. S(B) ≠ S(C); that is, sums of subsets cannot be equal.
2. If B contains more elements than C then S(B) > S(C).
Using sets.txt (a 4K text file with one-hundred sets containing seven to twelve
elements), identify all the special sum sets, A1, A2, …, Ak, and find the value
of S(A1) + S(A2) + ⋯ + S(Ak).
"""

import os
from pathlib import Path
from typing import List, Tuple


def parse_sets_from_file(filename: str) -> List[List[int]]:
    """Parse sets from file."""
    if not os.path.exists(filename):
        return []

    sets = []
    try:
        with open(filename, 'r') as f:
            for line in f:
                values = line.strip()
                if not values:
                    continue

                numbers = []
                for token in values.split(','):
                    try:
                        numbers.append(int(token))
                    except ValueError:
                        break
                else:
                    sets.append(sorted(numbers))
    except FileNotFoundError:
        return []

    return sets


def generate_subset_info(set_list: List[int]) -> List[Tuple[int, int, int]]:
    """Generate all non-empty subsets using bitmasks (efficient for small n ≤ 12).
    
    Returns array of [mask, sum, size] tuples.
    """
    n = len(set_list)
    subsets = []

    for mask in range(1, 1 << n):  # Skip empty subset (mask 0)
        sum_val = 0
        size = 0

        # Calculate sum and size for this mask
        for i in range(n):
            if (mask & (1 << i)) != 0:
                sum_val += set_list[i]
                size += 1

        subsets.append((mask, sum_val, size))

    return subsets


def is_special_sum_set(set_list: List[int]) -> bool:
    """Check if a set is a special sum set."""
    if not set_list:
        return False

    subsets = generate_subset_info(set_list)

    # Use pairwise comparison to enforce both special sum conditions
    for idx in range(len(subsets)):
        mask1, sum1, size1 = subsets[idx]

        for j in range(idx + 1, len(subsets)):
            mask2, sum2, size2 = subsets[j]
            if (mask1 & mask2) != 0:
                continue

            if sum1 == sum2:
                return False

            if size1 > size2:
                if sum1 <= sum2:
                    return False
            elif size2 > size1:
                if sum2 <= sum1:
                    return False

    return True


def main(filename: str = None) -> int:
    """Main function."""
    if filename is None:
        # Try to find sets.txt in common locations
        script_dir = Path(__file__).parent
        possible_paths = [
            script_dir.parent / 'data' / 'sets.txt',
            script_dir.parent / 'solutions' / 'sets.txt',
            script_dir / 'sets.txt',
        ]
        filename = None
        for path in possible_paths:
            if path.exists():
                filename = str(path)
                break

        if filename is None:
            return 0

    sets = parse_sets_from_file(filename)
    return sum(sum(set_list) if is_special_sum_set(set_list) else 0 for set_list in sets)


if __name__ == "__main__":
    print(main())

File: c/105.c
/* Project Euler Problem 105: Special Sum Sets - Testing */
#include <stdio.h>
#include <stdlib.h>
#include <stdbool.h>
#include <string.h>

static bool is_special_sum_set(int *set_list, int n) {
    int total_subsets = 1 << n;

    /* Precompute subset sums and sizes */
    int *sums = (int *)malloc(total_subsets * sizeof(int));
    int *sizes = (int *)malloc(total_subsets * sizeof(int));

    for (int mask = 0; mask < total_subsets; mask++) {
        int s = 0, sz = 0;
        for (int i = 0; i < n; i++) {
            if (mask & (1 << i)) {
                s += set_list[i];
                sz++;
            }
        }
        sums[mask] = s;
        sizes[mask] = sz;
    }

    bool result = true;
    for (int i = 1; i < total_subsets && result; i++) {
        for (int j = i + 1; j < total_subsets && result; j++) {
            if (i & j) continue; /* not disjoint */

            if (sums[i] == sums[j]) { result = false; break; }
            if (sizes[i] > sizes[j] && sums[i] <= sums[j]) { result = false; break; }
            if (sizes[j] > sizes[i] && sums[j] <= sums[i]) { result = false; break; }
        }
    }

    free(sums);
    free(sizes);
    return result;
}

int main(void) {
    FILE *f = fopen("solutions/sets.txt", "r");
    if (!f) f = fopen("data/sets.txt", "r");
    if (!f) f = fopen("../data/sets.txt", "r");
    if (!f) f = fopen("../solutions/sets.txt", "r");
    if (!f) { printf("0\n"); return 1; }

    int total_sum = 0;
    char line[1024];
    while (fgets(line, sizeof(line), f)) {
        int nums[20], n = 0;
        char *tok = strtok(line, ",\n\r ");
        while (tok && n < 20) {
            nums[n++] = atoi(tok);
            tok = strtok(NULL, ",\n\r ");
        }
        if (n == 0) continue;

        /* Sort */
        for (int i = 0; i < n - 1; i++)
            for (int j = i + 1; j < n; j++)
                if (nums[i] > nums[j]) { int t = nums[i]; nums[i] = nums[j]; nums[j] = t; }

        if (is_special_sum_set(nums, n)) {
            int s = 0;
            for (int i = 0; i < n; i++) s += nums[i];
            total_sum += s;
        }
    }
    fclose(f);

    printf("%d\n", total_sum);
    return 0;
}
